File: templates/problems/06_linked_list/05_remove_nth_from_end/python/solution.py
from helpers import ListNode, Optional


def solve_remove_nth_from_end(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    """Remove the nth node from the end using two pointers.

    Time: O(n), Space: O(1).
    """
    dummy = ListNode(0, head)
    fast: Optional[ListNode] = dummy
    slow: Optional[ListNode] = dummy
    for _ in range(n + 1):
        fast = fast.next  # type: ignore[union-attr]
    while fast is not None:
        fast = fast.next
        slow = slow.next  # type: ignore[union-attr]
    slow.next = slow.next.next  # type: ignore[union-attr]
    return dummy.next

File: templates/problems/06_linked_list/05_remove_nth_from_end/typescript/solution.ts
import { ListNode } from "./helpers";

export function solveRemoveNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    const dummy = new ListNode(0, head);
    let fast: ListNode | null = dummy;
    let slow: ListNode | null = dummy;
    for (let i = 0; i <= n; i++) {
        fast = fast!.next;
    }
    while (fast !== null) {
        fast = fast.next;
        slow = slow!.next;
    }
    slow!.next = slow!.next!.next;
    return dummy.next;
}
